File: 08/a/main.py
import itertools
import re
import sys

nodes = {}
current = "AAA"

LINE_REGEX = re.compile(r"([A-Z]{3}) = \(([A-Z]{3}), ([A-Z]{3})\)")

with open(sys.argv[1]) as f:
    for line in f:
        if line == "\n":
            continue
        elif "=" in line:
            (info,) = LINE_REGEX.findall(line)
            nodes[info[0]] = (info[1], info[2])
        else:
            instructions = line.strip()

for i, direction in enumerate(itertools.cycle(instructions)):
    if current == "ZZZ":
        break
    current = nodes[current][int(direction == "R")]

print(i)

File: 08/a/main.cpp
#include <fstream>
#include <iostream>
#include <regex>
#include <string>
#include <unordered_map>

int main(const int, const char** argv) {
  std::unordered_map<std::string, std::pair<std::string, std::string>> nodes;
  std::ifstream input_file_stream{argv[1]};
  std::string line, instructions, current{"AAA"};
  std::regex r{R"(([A-Z]+) = \(([A-Z]+), ([A-Z]+)\))"};
  std::smatch match;
  unsigned long long i{};

  while (std::getline(input_file_stream, line)) {
    if (line.empty()) {
      continue;
    } else if (line.contains('=')) {
      regex_search(line, match, r);
      nodes[match.str(1)] = std::make_pair(match.str(2), match.str(3));
    } else {
      instructions = line;
    }
  }
  for (; ; ++i) {
    if (current == "ZZZ") {
      break;
    }
    current = instructions[i % instructions.size()] == 'L' ? nodes.at(current).first : nodes.at(current).second;
  }

  std::cout << i << '\n';
}
